File: others/分类/二叉树/1026. 节点与其祖先之间的最大差值.cpp
#include <algorithm>
#include <queue>
#include <map>
#include <functional>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

using namespace std;

class Solution {
public:
    int mx = 0, mn = 0;
    int maxAncestorDiff(TreeNode *root) {
        if (!root) {
            return 0;
        }



        return helper(root, root->val, root->val);
    }

    int helper(TreeNode *node, int mx, int mn) {
        if (!node) {
            return mx - mn;
        }

        mx = max(mx, node->val);
        mn = min(mn, node->val);

        return max(helper(node->left, mx, mn), helper(node->right, mx, mn));
    }
};

File: leetcode/1000-1100/1026. 节点与其祖先之间的最大差值.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        ans = 0

        def dfs(root: Optional[TreeNode]) -> list:
            x = root.val
            nonlocal ans
            mn, mx = x, x
            if root.left:
                l = dfs(root.left)
                mn = min(mn, l[0])
                mx = max(mx, l[1])
                ans = max(ans, abs(x - l[0]), abs(x - l[1]))
            if root.right:
                r = dfs(root.right)
                mn = min(mn, r[0])
                mx = max(mx, r[1])
                ans = max(ans, abs(x - r[0]), abs(x - r[1]))
            return [mn, mx]

        dfs(root)
        return ans
